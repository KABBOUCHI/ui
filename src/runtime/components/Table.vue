<script lang="ts">
import { tv } from 'tailwind-variants'
import type { AppConfig } from '@nuxt/schema'
import type { ColumnDef, ColumnFiltersState, ExpandedState, SortingState, VisibilityState, Updater } from '@tanstack/vue-table'
import _appConfig from '#build/app.config'
import theme from '#build/ui/table'

const appConfig = _appConfig as AppConfig & { ui: { table: Partial<typeof theme> } }

const table = tv({ extend: tv(theme), ...(appConfig.ui?.table || {}) })

export interface TableProps<T> {
  columns?: ColumnDef<T>[]
  data: T[]
  class?: any
  ui?: Partial<typeof table.slots>
}

export type TableEmits = {}

export interface TableSlots<T> {
  expanded: { row: T }
}
</script>

<script setup lang="ts" generic="T">
import { ref, computed, type Ref } from 'vue'
import { FlexRender, getCoreRowModel, getExpandedRowModel, getFilteredRowModel, getPaginationRowModel, getSortedRowModel, useVueTable } from '@tanstack/vue-table'
import { upperFirst } from 'scule'

const props = defineProps<TableProps<T>>()
defineEmits<TableEmits>()
defineSlots<TableSlots<T>>()

// eslint-disable-next-line vue/no-dupe-keys
const ui = table()

const sorting = ref<SortingState>([])
const columnFilters = ref<ColumnFiltersState>([])
const columnVisibility = ref<VisibilityState>({})
const rowSelection = ref({})
const expanded = ref<ExpandedState>({})

const columns = computed<ColumnDef<T>[]>(() => props.columns ?? Object.keys(props.data[0] ?? {}).map((accessorKey: string) => ({ accessorKey, header: upperFirst(accessorKey) })))

const tableApi = useVueTable({
  data: props.data,
  columns: columns.value,
  getCoreRowModel: getCoreRowModel(),
  getPaginationRowModel: getPaginationRowModel(),
  getSortedRowModel: getSortedRowModel(),
  getFilteredRowModel: getFilteredRowModel(),
  getExpandedRowModel: getExpandedRowModel(),
  onSortingChange: updaterOrValue => valueUpdater(updaterOrValue, sorting),
  onColumnFiltersChange: updaterOrValue => valueUpdater(updaterOrValue, columnFilters),
  onColumnVisibilityChange: updaterOrValue => valueUpdater(updaterOrValue, columnVisibility),
  onRowSelectionChange: updaterOrValue => valueUpdater(updaterOrValue, rowSelection),
  onExpandedChange: updaterOrValue => valueUpdater(updaterOrValue, expanded),
  state: {
    get sorting() { return sorting.value },
    get columnFilters() { return columnFilters.value },
    get columnVisibility() { return columnVisibility.value },
    get rowSelection() { return rowSelection.value },
    get expanded() { return expanded.value }
  }
})

function valueUpdater<T extends Updater<any>>(updaterOrValue: T, ref: Ref) {
  ref.value = typeof updaterOrValue === 'function' ? updaterOrValue(ref.value) : updaterOrValue
}

defineExpose({
  tableApi
})
</script>

<template>
  <div :class="ui.root({ class: [props.class, props.ui?.root] })">
    <table :class="ui.base({ class: [props.ui?.base] })">
      <thead :class="ui.thead({ class: [props.ui?.thead] })">
        <tr v-for="headerGroup in tableApi.getHeaderGroups()" :key="headerGroup.id" :class="ui.tr({ class: [props.ui?.tr] })">
          <th v-for="header in headerGroup.headers" :key="header.id" :class="ui.th({ class: [props.ui?.th] })">
            <FlexRender v-if="!header.isPlaceholder" :render="header.column.columnDef.header" :props="header.getContext()" />
          </th>
        </tr>
      </thead>
      <tbody :class="ui.tbody({ class: [props.ui?.tbody] })">
        <template v-if="tableApi.getRowModel().rows?.length">
          <template v-for="row in tableApi.getRowModel().rows" :key="row.id">
            <tr :data-state="row.getIsSelected() && 'selected'" :class="ui.tr({ class: [props.ui?.tr] })">
              <td v-for="cell in row.getVisibleCells()" :key="cell.id" :class="ui.td({ class: [props.ui?.td] })">
                <FlexRender :render="cell.column.columnDef.cell" :props="cell.getContext()" />
              </td>
            </tr>
            <tr v-if="row.getIsExpanded()" :class="ui.tr({ class: [props.ui?.tr] })">
              <td :colspan="row.getAllCells().length" :class="ui.td({ class: [props.ui?.td] })">
                <slot name="expanded" :row="row" />
              </td>
            </tr>
          </template>
        </template>

        <tr v-else :class="ui.tr({ class: [props.ui?.tr] })">
          <td
            :colspan="columns?.length"
            class="h-24 text-center"
          >
            No results.
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</template>
