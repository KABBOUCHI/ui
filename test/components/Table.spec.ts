import { describe, it, expect } from 'vitest'
import Table, { type TableProps, type TableSlots } from '../../src/runtime/components/Table.vue'
import ComponentRender from '../component-render'

describe('Table', () => {
  const data = [{
    id: 'm5gr84i9',
    amount: 316,
    status: 'success',
    email: 'ken99@yahoo.com'
  }, {
    id: '3u1reuv4',
    amount: 242,
    status: 'success',
    email: 'Abe45@gmail.com'
  }, {
    id: 'derv1ws0',
    amount: 837,
    status: 'processing',
    email: 'Monserrat44@gmail.com'
  }, {
    id: '5kma53ae',
    amount: 874,
    status: 'success',
    email: 'Silas22@gmail.com'
  }, {
    id: 'bhqecj4p',
    amount: 721,
    status: 'failed',
    email: 'carmella@hotmail.com'
  }]

  const props = { data }

  it.each([
    // Props
    ['with data', { props }],
    ['with class', { props: { ...props, class: 'absolute' } }],
    ['with ui', { props: { ...props, ui: { base: 'table-auto' } } }]
  ])('renders %s correctly', async (nameOrHtml: string, options: { props?: TableProps<typeof data[number]>, slots?: Partial<TableSlots<typeof data[number]>> }) => {
    const html = await ComponentRender(nameOrHtml, options, Table)
    expect(html).toMatchSnapshot()
  })
})
