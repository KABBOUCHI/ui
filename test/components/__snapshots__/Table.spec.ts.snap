// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Table > renders with class correctly 1`] = `
"<div class="overflow-x-auto absolute">
  <table class="min-w-full table-fixed divide-y divide-[--ui-border-accented]">
    <thead class="">
      <tr class="">
        <th class="px-4 py-3.5 text-sm text-[--ui-text-highlighted] text-left rtl:text-right font-semibold">Id</th>
        <th class="px-4 py-3.5 text-sm text-[--ui-text-highlighted] text-left rtl:text-right font-semibold">Amount</th>
        <th class="px-4 py-3.5 text-sm text-[--ui-text-highlighted] text-left rtl:text-right font-semibold">Status</th>
        <th class="px-4 py-3.5 text-sm text-[--ui-text-highlighted] text-left rtl:text-right font-semibold">Email</th>
      </tr>
    </thead>
    <tbody class="divide-y divide-[--ui-border]">
      <tr class="">
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">m5gr84i9</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">316</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">success</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">ken99@yahoo.com</td>
      </tr>
      <!--v-if-->
      <tr class="">
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">3u1reuv4</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">242</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">success</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">Abe45@gmail.com</td>
      </tr>
      <!--v-if-->
      <tr class="">
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">derv1ws0</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">837</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">processing</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">Monserrat44@gmail.com</td>
      </tr>
      <!--v-if-->
      <tr class="">
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">5kma53ae</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">874</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">success</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">Silas22@gmail.com</td>
      </tr>
      <!--v-if-->
      <tr class="">
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">bhqecj4p</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">721</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">failed</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">carmella@hotmail.com</td>
      </tr>
      <!--v-if-->
    </tbody>
  </table>
</div>"
`;

exports[`Table > renders with data correctly 1`] = `
"<div class="relative overflow-x-auto">
  <table class="min-w-full table-fixed divide-y divide-[--ui-border-accented]">
    <thead class="">
      <tr class="">
        <th class="px-4 py-3.5 text-sm text-[--ui-text-highlighted] text-left rtl:text-right font-semibold">Id</th>
        <th class="px-4 py-3.5 text-sm text-[--ui-text-highlighted] text-left rtl:text-right font-semibold">Amount</th>
        <th class="px-4 py-3.5 text-sm text-[--ui-text-highlighted] text-left rtl:text-right font-semibold">Status</th>
        <th class="px-4 py-3.5 text-sm text-[--ui-text-highlighted] text-left rtl:text-right font-semibold">Email</th>
      </tr>
    </thead>
    <tbody class="divide-y divide-[--ui-border]">
      <tr class="">
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">m5gr84i9</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">316</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">success</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">ken99@yahoo.com</td>
      </tr>
      <!--v-if-->
      <tr class="">
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">3u1reuv4</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">242</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">success</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">Abe45@gmail.com</td>
      </tr>
      <!--v-if-->
      <tr class="">
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">derv1ws0</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">837</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">processing</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">Monserrat44@gmail.com</td>
      </tr>
      <!--v-if-->
      <tr class="">
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">5kma53ae</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">874</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">success</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">Silas22@gmail.com</td>
      </tr>
      <!--v-if-->
      <tr class="">
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">bhqecj4p</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">721</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">failed</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">carmella@hotmail.com</td>
      </tr>
      <!--v-if-->
    </tbody>
  </table>
</div>"
`;

exports[`Table > renders with ui correctly 1`] = `
"<div class="relative overflow-x-auto">
  <table class="min-w-full divide-y divide-[--ui-border-accented] table-auto">
    <thead class="">
      <tr class="">
        <th class="px-4 py-3.5 text-sm text-[--ui-text-highlighted] text-left rtl:text-right font-semibold">Id</th>
        <th class="px-4 py-3.5 text-sm text-[--ui-text-highlighted] text-left rtl:text-right font-semibold">Amount</th>
        <th class="px-4 py-3.5 text-sm text-[--ui-text-highlighted] text-left rtl:text-right font-semibold">Status</th>
        <th class="px-4 py-3.5 text-sm text-[--ui-text-highlighted] text-left rtl:text-right font-semibold">Email</th>
      </tr>
    </thead>
    <tbody class="divide-y divide-[--ui-border]">
      <tr class="">
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">m5gr84i9</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">316</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">success</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">ken99@yahoo.com</td>
      </tr>
      <!--v-if-->
      <tr class="">
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">3u1reuv4</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">242</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">success</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">Abe45@gmail.com</td>
      </tr>
      <!--v-if-->
      <tr class="">
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">derv1ws0</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">837</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">processing</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">Monserrat44@gmail.com</td>
      </tr>
      <!--v-if-->
      <tr class="">
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">5kma53ae</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">874</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">success</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">Silas22@gmail.com</td>
      </tr>
      <!--v-if-->
      <tr class="">
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">bhqecj4p</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">721</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">failed</td>
        <td class="p-4 text-sm text-[--ui-text-muted] whitespace-nowrap">carmella@hotmail.com</td>
      </tr>
      <!--v-if-->
    </tbody>
  </table>
</div>"
`;
